guides/micronaut-streamed-file-and-reactor-streaming-http-client/java/src/main/java/example/micronaut/HomeController.java

```diff
@@ -39,44 +34,14 @@ class HomeController implements AutoCloseable {
         this.reactorStreamingHttpClient = ReactorStreamingHttpClient.create(url); // <2>
     }
 
-    @Get("{?q}") // <3>
-    StreamedFile download(@Nullable @QueryValue String q) throws IOException { // <4>
-        URI uri = q == null ? DEFAULT_URI : URI.create(q);
-        HttpRequest<?> request = HttpRequest.GET(uri);
-
-        LOG.trace("downloading: {}", uri);
-        PipedOutputStream outputStream = new PipedOutputStream();
-        dataStreamToOutputStream(request, outputStream, () -> LOG.trace("finished download"));
-
-        PipedInputStream inputStream = new PipedInputStream(1024*10);
-        inputStream.connect(outputStream);
-
-        return new StreamedFile(inputStream, MediaType.IMAGE_PNG_TYPE);
-    }
-
-    private void dataStreamToOutputStream(HttpRequest<?> request,
-                                          PipedOutputStream outputStream,
-                                          Runnable finallyRunnable) {
-        reactorStreamingHttpClient.dataStream(request) // <5>
-                .doOnNext(byteBuffer -> {
-                    LOG.trace("Saving byte array");
-                    try {
-                        outputStream.write(byteBuffer.toByteArray());
-                    } catch (IOException e) {
-                        LOG.error("IO Exception closing the stream", e);
-                    }
-                })
-                .doFinally(signalType -> {
-                    try {
-                        LOG.trace("Closing OutputStream");
-                        outputStream.close();
-                    } catch (IOException e) {
-                        LOG.error("IO Exception closing the stream", e);
-                        System.out.println();
-                    }
-                    finallyRunnable.run();
-                })
-                .subscribe();
+    @Get // <3>
+    Flux<ByteBuffer<?>> download() {
+        HttpRequest<?> request = HttpRequest.GET(DEFAULT_URI);
+        return reactorStreamingHttpClient.dataStream(request).doOnNext(bb -> {
+            if (bb instanceof ReferenceCounted rc) {
+                rc.retain();
+            }
```

Comment:
the content will need to be updated to explain this new code

Reply:
for this code, we have a callout which says: 

> Request a stream of data where each emitted item is a ByteBuffer instance.

what do you think we should say instead?

Reply:
ping @graemerocher

Reply:
Something like:

```
The `dataStream` method emits instances of `ByteBuffer`. These instances are garbage collected after each chunk is emitted. In order to propagate the chunks to the response stream they need to be retained by calling the `retain()` method. The framework will call `release()` on these garbage collected chunks once each one is written to the response.
```